[#32] 코드 개선 및 테스트 추가

canlion opened the pull request:
* 누락된 기능 추가
  - GeoCoding 서비스 관련 예외 처리
* 테스트 추가
  - 레포지터리 관련 테스트 개선
    - @SpringBootTest 어노테이션을 사용하여 다른 요소 의한 문제로 레포지터리 테스트가 실행되지 못하는 문제
  - GeoCoding 서비스 관련 테스트 추가

canlion committed:
test: 레포지터리 테스트에 @DataJpaTest 적용

* @SpringBootTest 대신 @DataJpaTest를 적용하여 컨텍스트 로드 간소화
  - @SpringBootTest는 어플리케이션 전체 컨텍스트를 로드하여 테스트와
    무관한 컴포넌트 문제로 테스트가 실패하며 문제를 찾기 어렵게 만듦.
* 일부 테스트 제거
  - JPA 레포지터리에 엔티티를 전달하는 단순 CRUD에 대한 테스트를
    유지해야하는지에 대한 고민 중.
  - 남아있는 테스트는 테스트 코드 작성 형태를 보이기 위해 유지.

canlion committed:
refactor: geo-coding 서비스에서 geo-coding 요청 클라이언트를 분리

* geo-coding API를 호출하는 클라이언트를 서비스에서 분리
  - 테스트 코드에서 mocking 대상을 명시적으로 제시함.
  - 네이버 클라우드 외의 geo-coding API를 염두에 둠.
    - 단, 현재 geo-coding 서비스의 로직이 네이버 클라우드에 종속된
      상태임.

canlion committed:
feat: geo-coding 결과가 없는 경우 예외 발생시킴

* 주어진 주소에 대한 geo-coding 결과가 없는 경우 InvalidAddressException
  발생시킴

canlion committed:
test: geoCodingService 테스트 추가

canlion commented:
## 커밋 상세
### [test: 레포지터리 테스트에 @DataJpaTest 적용](https://github.com/f-lab-edu/burger-map/pull/33/commits/c5fe962c029919195f8a4f2405d12db510cbef69)

* `@SpringBootTest`를 제거하고 `@DataJpaTest`로 대체
  - 레포지터리 테스트가 `@SpringBootTest` 어노테이션을 적용하여 테스트가 다른 요소에 영향을 받음.
    - `@SpringBootTest`에 의해 어플리케이션 전체 컨텍스트가 로드될 때 외부 프로퍼티 값을 요구하는 외부 API 호출 서비스가 프로퍼티 값의 부재 등으로 빈 생성이 안되면서 테스트 전체가 실패함
  - 테스트 대상과 전혀 무관한 요소의 문제로 테스트 자체가 진행이 되지 않는 상황은 문제점을 찾기 어렵게 만든다고 생각되어 `@SpringBootTest`를 제거하고 `@DataJpaTest`로 대체
    - `@DataJpaTest`: JPA 관련 컨텍스트만을 로드하고 `TestEntityManager` 제공. `@Repository`를 스캔하지는 않아 테스트에 `@Import`를 이용해 테스트 대상인 레포지터리를 명시적으로 임포트함. 
* 레포지터리가 단순 CRUD만을 수행하므로 테스트의 필요성에 비해 실행시간이 길다고 생각되어 대부분 제거
  - 필요시에 다시 작성할 수 있도록 샘플 코드 형식으로 일부만 유지

### [refactor: geo-coding 서비스에서 geo-coding 요청 클라이언트를 분리](https://github.com/f-lab-edu/burger-map/pull/33/commits/accc76b9223c26e98ca9209796ba109c225abf1b)

* 테스트 코드 작성이 쉽도록 서비스에 포함된 `WebClient`를 별도의 클래스로 분리
  - 테스트 코드 작성시 mocking 대상을 명시적으로 제시
* `@Value`를 `@ConfigurationProperties`로 대체
  - `@ConfigurationProperties`는 관련된 다수의 값 주입에 유리
  - 유연한 속성-프로퍼티 매칭 제공

## 질문

* 외부 API 호출을 위한 API key 문제로 401 예외가 발생하는 경우에 대한 테스트 코드가 필요할까요?
  - 로직문제가 아닌 오타와 비슷한 개발자의 단순 실수로 생각되어 딱히 필요하지는 않다는 생각이 듭니다.
  - 통합 테스트에서 잡을 수 있는 문제라고 생각됩니다.

f-lab-lyan commented on api/src/main/java/burgermap/service/geoCoding/NCPGeoCodingClient.java:
```diff
@@ -0,0 +1,39 @@
+
+    @PostConstruct
+    void init() {
+        System.out.println("clientId = " + clientId);
```
```suggestion
        logger.debug("clientId = " + clientId);
```

f-lab-lyan reviewed (approved):
수고하셨습니다. :-)

canlion committed:
refactor: GeoCodingClient 인터페이스 제거

* 불필요한 인터페이스 제거
  - 당분간 다른 클라우드의 geo-coding 서비스 추가 계획 없음
  - GeoCodingService가 NCP의 geo-coding에 의존적

canlion committed:
refactor: ImageService의 구성 요소를 역할에 따라 별도 클래스로 분리

* ImageService의 S3Presigner와 이미지업로드 이름 생성 메서드를 별도의
  클래스로 분리
  - 테스트 코드 작성시 원활하게 mocking 수행
  - 이미지 이름 생성 규칙 변경을 염두에 두고 인터페이스 도입

canlion committed:
test: ImageService 테스트 추가

canlion committed:
chore: 클래스명 변경

* 클래스 및 관련 참조에서 URL을 Url로 수정하여 네이밍 일관성을 유지

canlion committed:
test: StoreService 관련 테스트 추가

canlion committed:
refactor: filterFoods 메서드를 StoreService로 이동

* 단순 조회 기능보다는 비즈니스 로직을 담은 기능에 가까워 StoreService로 이동

canlion committed:
test: FoodService 테스트 추가

f-lab-lyan commented on api/src/main/java/burgermap/service/image/ImageService.java:
```diff
@@ -0,0 +1,71 @@
+        }
+
+        String imageUploadName = imageNameGenerator.generateImageName(fileName);
+        String imageUploadPath = String.join("/", imageUploadDirectory, imageUploadName);
```
Path를 사용해서 만들어 보시면 더 깔끔하지 않을까 싶네요.

f-lab-lyan replied:
https://www.baeldung.com/java-file-vs-file-path-separator
